From grinberg@compulab.co.il  Tue Oct 12 11:04:03 2010
From: Igor Grinberg <grinberg@compulab.co.il>
Date: Sun, 10 Oct 2010 17:59:18 +0200
Subject: USB: otg/ulpi: improve ulpi phy detection.
To: Greg Kroah-Hartman <gregkh@suse.de>
Cc: Daniel Mack <daniel@caiaq.de>, linux-usb@vger.kernel.org, Igor Grinberg <grinberg@compulab.co.il>
Message-ID: <1286726359-30349-2-git-send-email-grinberg@compulab.co.il>


Improve ulpi phy detection by utilizing the "scratch" register.
Allow unknown ulpi phy work without the need to hard-code the id.

Signed-off-by: Igor Grinberg <grinberg@compulab.co.il>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/otg/ulpi.c |   35 ++++++++++++++++++++++++++++++++---
 1 file changed, 32 insertions(+), 3 deletions(-)

--- a/drivers/usb/otg/ulpi.c
+++ b/drivers/usb/otg/ulpi.c
@@ -137,6 +137,32 @@ static int ulpi_set_flags(struct otg_tra
 	return ulpi_set_fc_flags(otg);
 }
 
+static int ulpi_check_integrity(struct otg_transceiver *otg)
+{
+	int ret, i;
+	unsigned int val = 0x55;
+
+	for (i = 0; i < 2; i++) {
+		ret = otg_io_write(otg, val, ULPI_SCRATCH);
+		if (ret < 0)
+			return ret;
+
+		ret = otg_io_read(otg, ULPI_SCRATCH);
+		if (ret < 0)
+			return ret;
+
+		if (ret != val) {
+			pr_err("ULPI integrity check: failed!");
+			return -ENODEV;
+		}
+		val = val << 1;
+	}
+
+	pr_info("ULPI integrity check: passed.\n");
+
+	return 0;
+}
+
 static int ulpi_init(struct otg_transceiver *otg)
 {
 	int i, vid, pid, ret;
@@ -155,10 +181,13 @@ static int ulpi_init(struct otg_transcei
 
 	for (i = 0; i < ARRAY_SIZE(ulpi_ids); i++)
 		if (ulpi_ids[i] == ULPI_ID(vid, pid))
-			return ulpi_set_flags(otg);
+			break;
+
+	ret = ulpi_check_integrity(otg);
+	if (ret)
+		return ret;
 
-	pr_err("ULPI ID does not match any known transceiver.\n");
-	return -ENODEV;
+	return ulpi_set_flags(otg);
 }
 
 static int ulpi_set_host(struct otg_transceiver *otg, struct usb_bus *host)
